src/pubtools/sign/clients/registry.py

```diff
@@ -0,0 +1,168 @@
+                )
+                break
+        else:
+            auth = (self.username, self.password)
+        return auth
+
+    def authenticate_to_registry(self, image_reference: str, auth_header: str) -> str:
+        """Ask for auth token based on given auth header.
+
+        Ask for authentication token based on the given auth header.
+
+        Args:
+            image_reference (str): Image reference to resolve authentication for.
+            auth_header (str): Authentication header from the registry.
+        Returns:
+            str: Authentication token.
+        """
+        _, _, value = auth_header.partition("Bearer")
+        items = parse_http_list(value)
+        opts = parse_keqv_list(items)
+        unparse_parts = [
+            "https",
+            opts["realm"].replace("https://", "").split("/", 1)[0],
+            opts["realm"].replace("https://", "").split("/", 1)[1],
+            "",
+            urlencode({"service": opts["service"], "scope": opts["scope"]}),
+            "",
+        ]
+        auth_url = urlunparse(unparse_parts)
+        username, password = self.resolve_authentication(image_reference)
+        response = self.session.get(auth_url, auth=(username, password))
+        if response.status_code == 200:
+            return response.json().get("token")
+        else:
+            response.raise_for_status()
+
+    def check_container_image_exists(
+        self, image_reference: str, auth_token: Optional[AuthTokenWrapper] = None
+    ) -> bool:
+        """Check if the given container image exists.
+
+        Args:
+            image_reference (str): Image reference to check.
+            auth_token (Optional[AuthTokenWrapper]): Authentication token.
+        Returns:
+            bool: True if the image exists, False otherwise.
+        """
+        repo_ref, tag = image_reference.rsplit(":", 1)
+        registry, repo = repo_ref.split("/", 1)
+        manifest_url = f"https://{registry}/v2/{repo}/manifests/{tag}"
+        headers = {"Authorization": f"Bearer {auth_token.token}"}
+        response = self.session.get(manifest_url, headers=headers)
+
+        if response.status_code == 200:
+            return True
+        elif response.status_code == 401:
+            auth_header = response.headers.get("www-authenticate")
+            auth_token.token = self.authenticate_to_registry(image_reference, auth_header)
+            # Retry the original request with the token
+            headers = {"Authorization": f"Bearer {auth_token.token}"}
+            response = self.session.get(manifest_url, headers=headers)
+            if response.status_code == 200:
+                return True
+            elif response.status_code == 404:
+                return False
+            else:
+                return False
+        elif response.status_code == 404:
+            return False
+        else:
+            print(f"Error: {response.status_code} - {response.text}", file=sys.stderr)
```

Comment:
Print. Also, Can't the conditions be simplified?

Reply:
How would you simplified it?